main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -33,303 +33,518 @@ object Safety {
     //
     // Collect all errors.
     //
-    val defErrors = root.defs.flatMap {
-      case (_, defn) => visitDef(defn, root)
-    }
 
-    val instanceErrors = visitSendable(root)
+    for {
+      _ <- Validation.foldRight(root.defs.toList)(root.toSuccess) {
+        case ((_, defn), _) => visitDef(defn, root)
+      }
 
-    val errors = defErrors ++ instanceErrors
+      errors <- visitSendable(root)
+    } yield errors
 
     //
     // Check if any errors were detected.
     //
-    if (errors.isEmpty)
-      root.toSuccess
-    else
-      Validation.Failure(errors.to(LazyList))
   }
 
   /**
     * Checks that no type parameters for types that implement `Sendable` of kind `Region`
     */
-  private def visitSendable(root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitSendable(root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = {
 
     val sendableClass = new Symbol.ClassSym(Nil, "Sendable", SourceLocation.Unknown)
 
-    root.instances.getOrElse(sendableClass, Nil) flatMap {
-      case Instance(_, _, _, _, tpe, _, _, _, loc) =>
-        if (tpe.typeArguments.exists(_.kind == Kind.Bool))
-          List(SafetyError.SendableError(tpe, loc))
-        else
-          Nil
+    Validation.foldRight(root.instances.getOrElse(sendableClass, Nil))(root.toSuccess[Root, CompilationMessage]) {
+      case (Instance(_, _, _, _, tpe, _, _, _, loc), _) =>
+        if (tpe.typeArguments.exists(_.kind == Kind.Bool)) {
+          SoftFailure[Root, CompilationMessage](root, LazyList(SafetyError.SendableError(tpe, loc)))
+        } else {
+          root.toSuccess[Root, CompilationMessage]
+        }
     }
   }
 
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = {
     val renv = def0.spec.tparams.map(_.sym).foldLeft(RigidityEnv.empty) {
       case (acc, e) => acc.markRigid(e)
     }
-    visitExp(def0.impl.exp, renv, root)
+    mapN(visitExp(def0.impl.exp, renv, root))(_ => root)
   }
 
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): Validation[Expression, CompilationMessage] = {
 
-    def visit(exp0: Expression): List[CompilationMessage] = exp0 match {
-      case Expression.Cst(_, _, _) => Nil
+    def visit(exp0: Expression): Validation[Expression, CompilationMessage] = exp0 match {
+      case Expression.Cst(_, _, _) => Success(exp0)
```

Comment:
Stylistically we almost always use the implicit method `toSuccess`

Reply:
Was also my original choice, but the compiler refused that without type annotation in a lot of cases.